el-patch-template.el

```diff
@@ -0,0 +1,554 @@
+resolution and `:matched' which is set to t if the template is
+matched to a form in DEFINITION."
+  (let (matched-forms-count matched-ptemplate)
+    (cl-dolist (ptemplate ptemplates)
+      (let ((matched (el-patch-template--match-p definition
+                                                 (plist-get ptemplate :old))))
+        (when matched
+          (when matched-ptemplate
+            (error "A form matches multiple templates"))
+          (setq matched-forms-count matched
+                matched-ptemplate ptemplate))))
+    (cond
+     ((null matched-ptemplate)
+      (if (consp definition)
+          (cons (el-patch-template--apply (car definition)
+                                          ptemplates)
+                (el-patch-template--apply (cdr definition)
+                                          ptemplates))
+        definition))
+     ((plist-get matched-ptemplate :matched)
+      (error "A template matches multiple forms"))
+     ((and (consp definition)
+           (or
+            (el-patch-template--any-p (car definition)
+                                      ptemplates)
+            (and
+             (cdr definition)
+             (el-patch-template--any-p (cdr definition)
+                                       ptemplates
+                                       (1- matched-forms-count)))))
+      (error "A form matching a template has subforms matching\
+ other templates"))
+     (t
+      ;; The old resolution of the template uniquely matches the definition
+      ;; Here we first mark the template as being matched then
+      ;; do the actual resolution
+      (plist-put matched-ptemplate :matched t)
+      (let ((resolution
+             (el-patch-template--process definition
+                                         (list
+                                          (plist-get matched-ptemplate
+                                                     :template)))))
+        (cons (caar resolution)
+              (el-patch-template--apply (cdr resolution)
+                                        ptemplates)))))))
+
+(defun el-patch-template--resolve (forms)
+  "Resolve `el-patch-*' directives in FORMS.
+
+Similar to `el-patch--resolve' with a special treatment for
+`el-patch-concat'. Specifically, if the arguments of
+`el-patch-concat' have `...' in them, it is not resolved but
+changed to `el-patch-template--concat'."
+  (cl-letf* ((old-concat (symbol-function 'concat))
+             ((symbol-function 'concat)
+              (lambda (&rest args)
+                (if (cl-some (lambda (x) (equal x '...)) args)
+                    (cons 'el-patch-template--concat args)
+                  (apply old-concat args)))))
+    (el-patch--resolve forms nil)))
+
+(defun el-patch-template--impl (keyword-name templates)
+  "Apply the templates to the definition of an elisp construct.
+
+The actual implementation of `el-patch-template', accepts the
+same arguments but quoted."
```

Comment:
Lint complains about the documentation here (missing argument names). Should I copy the documentation from `el-patch-template`? Seems redundant.

Reply:
Checkdoc is a bit silly sometimes. What I've done in this case is something like this:

```elisp
(defun el-patch-template--impl (keyword-name templates)
  "Apply the templates to the definition of an elisp construct.
The meaning of KEYWORD-NAME and TEMPLATES is the same as in
`el-patch-template' (which see), but here they need to be quoted
since they are passed as regular function arguments."
```